File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/routes/AdminController.java
package dev.tenantcore.auth.routes;

import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.Tenant;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.Map;

@RestController
@RequestMapping("/admin/tenants")
public class AdminController {

    private final InMemoryStore store;

    public AdminController(InMemoryStore store) {
        this.store = store;
    }

    @GetMapping
    @PreAuthorize("hasAuthority('admin:tenant:create')")
    public ResponseEntity<Collection<Tenant>> listTenants() {
        return ResponseEntity.ok(store.getAllTenants());
    }

    @PostMapping
    @PreAuthorize("hasAuthority('admin:tenant:create')")
    public ResponseEntity<?> createTenant(@RequestBody Map<String, String> body) {
        Tenant tenant = store.createTenant(
            body.getOrDefault("slug", "tenant-" + System.currentTimeMillis()),
            body.getOrDefault("name", "Tenant"),
            body.getOrDefault("first_admin_email", "admin@example.test")
        );
        return ResponseEntity.status(HttpStatus.CREATED).body(tenant);
    }
}

File: examples/canonical-auth/elixir/lib/tenantcore_auth/controllers/admin_controller.ex
defmodule TenantcoreAuth.AdminController do
  use TenantcoreAuth, :controller

  def index(%Plug.Conn{assigns: %{auth: auth}} = conn, _params) do
    if auth.tenant_role == "super_admin" and "admin:tenant:create" in auth.permissions do
      tenants =
        TenantcoreAuth.Domain.InMemoryStore.list_tenants()
        |> Enum.map(fn tenant -> %{id: tenant.id, slug: tenant.slug, name: tenant.name} end)

      json(conn, %{tenants: tenants})
    else
      conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end

  def create(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"slug" => slug, "name" => name, "first_admin_email" => email}) do
    if auth.tenant_role == "super_admin" and "admin:tenant:create" in auth.permissions do
      tenant = TenantcoreAuth.Domain.InMemoryStore.create_tenant(slug, name, email)
      conn |> put_status(:created) |> json(%{tenant: %{id: tenant.id, slug: tenant.slug, name: tenant.name}})
    else
      conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end
end
